src/main/java/com/ocient/jdbc/XGConnection.java

```
@@ -219,7 +219,7 @@ public void run()
 						throw e;
 					}
 
-					reconnectOrThrow();
+					blockingGetSQLResult(reconnect());
```

hmm, I'm not sure we want to swallow reconnect failures, I think what you had before was right...

Or `blockingGetSQLResult(reconnect()).get();` which is just syntactic sugar for your `reconnectOrThrow()` helper method.